specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-04-01-preview/services.json

```diff
@@ -0,0 +1,399 @@
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/ResourceGroupNameParameter"
+          },
+          {
+            "$ref": "#/parameters/accountNameParameter"
+          },
+          {
+            "$ref": "#/parameters/serviceNameParameter"
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/ApiVersionParameter"
+          }
+        ],
+        "responses": {
+          "202": {
+            "description": "The job will complete asynchronously."
+          },
+          "204": {
+            "description": "The response for service deletion."
+          },
+          "200": {
+            "description": "The response for service deletion."
+          }
+        }
+      }
+    }
+  },
+  "definitions": {
+    "ServicesCreateUpdateParameters": {
+      "description": "Service create or update parameters.",
+      "type": "object",
+      "properties": {
+        "properties": {
+          "type": "object",
+          "x-ms-client-flatten": true,
+          "description": "The properties for creating services.",
+          "$ref": "#/definitions/ServicesCreateUpdateProperties"
+        }
+      },
+      "required": [
+        "properties"
+      ],
+      "x-ms-azure-resource": true
+    },
+    "ServicesCreateUpdateProperties": {
+      "description": "Describes the service response property.",
+      "type": "object",
+      "properties": {
+        "instanceSize": {
+          "$ref": "#/definitions/ServiceResourceInstanceSize"
+        },
+        "instanceCount": {
+          "description": "Instance count for the service.",
+          "type": "integer"
+        },
+        "serviceType": {
+          "type": "string",
+          "description": "The type of Azure resource."
+        }
+      },
+      "required": [
+        "instanceSize",
+        "instanceCount"
+      ]
+    },
+    "ServiceResource": {
+      "description": "Services response resource.",
+      "type": "object",
+      "discriminator": "serviceType",
+      "allOf": [
+        {
+          "$ref": ".../../cosmos-db.json#/definitions/ARMProxyResource"
+        }
+      ],
+      "properties": {
+        "creationTime": {
+          "readOnly": true,
+          "type": "string",
+          "format": "date-time",
+          "description": "Time of the last state change (ISO-8601 format)."
+        },
+        "instanceSize": {
+          "$ref": "#/definitions/ServiceResourceInstanceSize"
+        },
+        "instanceCount": {
+          "description": "Instance count for the service.",
+          "type": "integer"
+        },
+        "serviceType": {
+          "type": "string",
+          "description": "The type of Azure resource."
+        },
+        "status": {
+          "$ref": "#/definitions/ServiceResourceStatus"
+        }
+      },
+      "additionalProperties": {
```

Comment:
>additionalProperties [](http://example.com/codeflow?start=7&length=20)

Why is this a loose schema? [https://armwiki.azurewebsites.net/api_contracts/guidelines/policy.html#plcy005-avoid-free-form-objects](https://armwiki.azurewebsites.net/api_contracts/guidelines/policy.html#plcy005-avoid-free-form-objects)

Reply:
So our Services APIs offer control over different types of Services. Today, it is DataTransfer and SqlDedicatedGateway. The base ServiceResource is same but they DataTransferResource and SqlDedicatedGatewayResource have different resource models. To capture this polymorphism, I did it this way.  (In future we want to add some more services like these)

I followed Azure Datafactory as an example: https://github.com/Azure/azure-rest-api-specs/blob/21c970568229eb388d0fce7097e889411ee0a85f/specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/entityTypes/Dataset.json#L75

Reply:
Please recommend if there is a better way to do this.

Reply:
Have you considered using a one of these options?
1. Splitting into multiple resource types?
2. Using the discriminator pattern to effectively choose 1 out of n schemas

Reply:
Platform services like Azure policy won't work on dynamic properties.

Reply:
@majastrz @j5lim can you please point me to an example to the recommended way of doing this

Reply:
> Have you considered using a one of these options?
> 
> 1. Splitting into multiple resource types?
> 2. Using the discriminator pattern to effectively choose 1 out of n schemas

Some questions: 
1. If I split into multiple resources, then how would I model the get, put APIs. What resource kind would they be returning ?? 

 2. Can you point to an example for 2. ?

Reply:
Since OpenAPI 2.0 doesnt support anyOf or oneOf, I have added the additionalProperties back to ServiceResource.